experiment/selinux/src/selinux.rs

```diff
@@ -0,0 +1,302 @@
+use std::sync::Once;
+use crate::xattr::*;
+use nix::unistd::gettid;
+use nix::sys::statfs;
+use nix::errno::Errno;
+use std::path::PathBuf;
+use std::fs::File;
+use std::io::{self, Read};
+use std::os::fd::{AsFd, AsRawFd};
+
+const XATTR_NAME_SELINUX: &str = "security.selinux";
+const ERR_EMPTY_PATH: &str = "empty path";
+static ATTR_PATH_INIT: Once = Once::new();
+static mut HAVE_THREAD_SELF: bool = false;
+
+pub fn set_disabled() {
+    panic!("not implemented yet")
+}
+
+pub fn get_enabled() -> bool {
+    panic!("not implemented yet")
+}
+
+pub fn class_index(class: &str) -> Result<i64, String> {
+    panic!("not implemented yet")
+}
+
+// set_file_label sets the SELinux label for this path, following symlinks, or returns an error.
+pub fn set_file_label(fpath: &str, label: &str) -> Result<(), std::io::Error> {
+    if fpath.is_empty() {
+        return Err(std::io::Error::new(io::ErrorKind::InvalidInput, ERR_EMPTY_PATH));
+    }
+
+    loop {
+        match set_xattr(fpath, XATTR_NAME_SELINUX, label.as_bytes(), 0) {
+            Ok(_) => break,
+// TODO            Err(Errno::EINTR) => continue,
+            Err(e) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("set_xattr failed: {}", e),
+                ));
+            }
+        }
+    }
+    Ok(())
+}
+
+pub fn lset_file_label(fpath: &str, label: &str) -> Result<(), std::io::Error> {
+    if fpath.is_empty() {
+        return Err(std::io::Error::new(io::ErrorKind::InvalidInput, ERR_EMPTY_PATH));
+    }
+
+    loop {
+        match lset_xattr(fpath, XATTR_NAME_SELINUX, label.as_bytes(), 0) {
+            Ok(_) => break,
+            // TODO: EINTR (after fixing lset_attr)
+            Err(e) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("lset_xattr failed: {}", e),
+                ));
+            }
+        }
+    }
+    Ok(())
+}
+
+pub fn file_label(fpath: &str) -> Result<String, std::io::Error> {
+    if fpath.is_empty() {
+        return Err(std::io::Error::new(io::ErrorKind::InvalidInput, ERR_EMPTY_PATH));
+    }
+    let label = get_xattr(fpath, XATTR_NAME_SELINUX);
+    match label {
+        Ok(mut v) => {            
+            if (v.len() > 0) && (v.chars().nth(v.len() - 1) == Some('\x00')) {
+                v = (&v[0..v.len() - 1]).to_string();
+            }
+            return Ok(v);
+        },
```

Comment:
as per the current definition, `get_xattr` already returns a `String` . So can we instead do this checking there (when it will be implemented in next PR) , and instead just return the label from here?

Also, instead of `len()` check, more idiomatic way would be `!v.is_empty()` , and for the  last element, we can do `v.chars().last()` . I also think there would be some other way to convert null terminated string into rust String; although the way will depend on how we are getting the original string in `get_xattr` implementation.

Reply:
> So can we instead do this checking there (when it will be implemented in next PR) , and instead just return the label from here?

Make sense, thanks. `get_xattr` is used in `file_label` and `lfile_label`. These two files handle label same. So we can move these logic to `get_attr`.